Make --limit-posts option faster

emad-elsaid opened the pull request:
<!--
  Thanks for creating a Pull Request! Before you submit, please make sure
  you've done the following:

  - I read the contributing document at https://jekyllrb.com/docs/contributing/
-->

<!--
  Make our lives easier! Choose one of the following by uncommenting it:
-->

<!-- This is a 🐛 bug fix. -->
This is a 🙋 feature or enhancement.
<!-- This is a 🔦 documentation change. -->
<!-- This is a 🔨 code refactoring. -->

<!--
  Before you submit this pull request, make sure to have a look at the following
  checklist. If you don't know how to do some of these, that's fine! Submit
  your pull request and we will help you out on the way.

  - I've added tests (if it's a bug, feature or enhancement)
  - I've adjusted the documentation (if it's a feature or enhancement)
  - The test suite passes locally (run `script/cibuild` to verify this)
-->

## Summary

When `_posts` directory has too many files (mine 60k). running `jekyll server --limit-posts 10` will take around 16 seconds, I expected limiting the posts to makes it take shorter time.

After digging into the code I found that the post reader reads all files all the time, then limit them, that takes too long and we don't need to read more than the limit.

So my solution was to pass down the limit to the part that reads the posts from the directory and limit that.

This makes `jekyll server --limit-posts 10` in the same directory takes 3 seconds that's 5.3x faster.

ashmaroli commented on lib/jekyll/reader.rb:
```diff
@@ -131,12 +132,15 @@ def filter_entries(entries, base_directory = nil)
       entries = Dir.chdir(base) { filter_entries(Dir["**/*"], base) }
       entries.delete_if { |e| File.directory?(site.in_source_dir(base, e)) }
+
+      entries_limit = [entries.length, limit].min
```
The check for positive `limit` in the next line can be hoisted to enclose this line as well and avoid allocating an array when `limit == 0`.

emad-elsaid replied:
to avoid this allocation I'll use trinary operator I see it's used here https://github.com/jekyll/jekyll/blob/master/lib/jekyll/site.rb#L480

ashmaroli replied:
Fair enough.

ashmaroli reviewed (changes requested):
Overall, very good first contribution, @emad-elsaid!
However, there are some issues that need to be addressed, notably inconsistencies with making the new parameter optional.

Also, I don't see a value in exposing `:limit` as a *public* getter method especially since it is the same as `site.limit_posts` . Let us have it as a `private` `attr_reader` for now and upgrade in future if requested for.

Additionally, this needs some tests to confirm nothing's broken. You're free to choose between Cucumber based tests and Minitest based tests.

ashmaroli commented on lib/jekyll/reader.rb:
```diff
@@ -131,12 +132,15 @@ def filter_entries(entries, base_directory = nil)
       entries.delete_if { |e| File.directory?(site.in_source_dir(base, e)) }
+
+      entries_limit = [entries.length, limit].min
+      limit.positive? ? entries[-entries_limit, entries_limit] : entries
```
I believe this line can be refactored to use `:last(n)`:
```suggestion
      limit.positive? ? entries.last(entries_limit) : entries
```
But, do benchmark the two routes and see if there's a benefit from making the change..

emad-elsaid commented:
I'm not sure about the tests, while developing this some tests covered the `--limit-posts` and made sure the correct posts are in the result https://github.com/jekyll/jekyll/blob/master/features/site_configuration.feature#L294 

and other tests where using the classes I changed, which complained about the new argument so I added the default value for it.

So the case I'm covering here is reducing the time when there are large number of posts while using `--limit-posts` so I'm not sure how to ensure this with a test so it wouldn't regress.

ashmaroli commented:
Okay. I've confirmed that the test coverage is already adequate. So, getting a pass status from CI will suffice here.

ashmaroli commented:
The following line
https://github.com/jekyll/jekyll/blob/3e3d9bce64d576b8d59271548fb4c2e280b93658/lib/jekyll/site.rb#L182

needs to be removed now since the limiting is handled as part of the `site.reader.read` call..

emad-elsaid commented:
I think if I removed `limit_posts!` this will happen:
- a site with 10 posts and 10 drafts
- limit number of posts to 5
- the reader will limit posts to last 5 and limit drafts to 5
- return 10 posts+drafts instead of total 5
so what current code does is after returning the limited files per directory it limit the sum of them
when I tried to remove it a test failed: 
```
Failure:
TestGeneratedSite#test_: generating limited posts should generate only the specified number of posts.  [/........../jekyll/test/test_generated_site.rb:92]
Minitest::Assertion: Expected: 5
  Actual: 8

914 tests, 2023 assertions, 1 failures, 0 errors, 0 skips
```

one other thing, there is a test in travis that fails but when I run it on my machine it passes, so not sure why this is happening
```
cucumber features/site_configuration.feature:294 # Scenario: Limit the number of posts generated by most recent date
```

ashmaroli commented:
Drafts and posts are documents of the same collection (`posts`), so they'll end up within `site.posts.docs`.
However, computing the limit and filtering twice does not seem ideal to me.

ashmaroli commented:
> there is a test in travis that fails but when I run it on my machine it passes

Not sure about this. It passes successfully at my end as well and on AppVeyor CI..
I've restarted the failing jobs on Travis to confirm the behavior.

ashmaroli commented:
After much thought, I've come to the conclusion that this change is not backwards-compatible.
It is a change in behavior that is markedly different from the long-standing documented feature.

Perhaps in the next major version..

emad-elsaid commented:
> Drafts and posts are documents of the same collection (`posts`), so they'll end up within `site.posts.docs`.
> However, computing the limit and filtering twice does not seem ideal to me.

Yeah in this case then the current implementation to limit the number of posts in the method that returns posts in a directory needs to be improved, I'll try to find another after we read and merge all directories then limit it and before reading the content of the files themselves.

shushugah commented on lib/jekyll/reader.rb:
```diff
@@ -131,16 +132,21 @@ def filter_entries(entries, base_directory = nil)
       entries = Dir.chdir(base) { filter_entries(Dir["**/*"], base) }
       entries.delete_if { |e| File.directory?(site.in_source_dir(base, e)) }
+
+      entries_limit = entries.length < limit ? entries.length : limit
```
this can be rewritten has `entries_limit = [entries.length, limit].min`
`

ashmaroli replied:
@shushugah The use of `Array#min` was suggested to be replaced in a previous review to avoid unnecessary allocations.

emad-elsaid commented:
it's been here and I don't have time to get back to it. so I'll close this.
